Fullscreen Layout Redesign

Allcharles opened the pull request:
# Fullscreen layout redesign

This PR updates the fullscreen layouts to give users access to the secondary and action menus.

## Changes

- Added menu toggle to full-screen page headers
- Added side-nav component
- Extracted secondary and action menus from router
- Extracted header links to `PrimaryMenuComponent`
- Updated styling of `CardsComponent`

## Problems

- Need to complete tests after architecture approval

## Issues

Closes #458 

## Visual Changes

### Menu System

![Menu System](https://user-images.githubusercontent.com/3955116/150075420-1eb815c1-fb7b-4ada-a3fd-7545fb894ca9.gif)

### Card resizing

![Cards](https://user-images.githubusercontent.com/3955116/150076098-777822c8-63a5-484b-b828-caaf1066fbc0.gif)

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

atruskie commented on src/app/app.component.html:
```diff
@@ -1,24 +1,45 @@
+  >
+    <ng-container *ngIf="fullscreen; else menuLayout">
+      <div id="menu">
+        <baw-off-canvas>
```
given `canvas` is an actual concept (dom element, JS API etc...) i'd suggest this name is bad

Maybe `baw-modal` or `baw-dynamic-menu`? I'll look at the actual component definition later and maybe come up with a better name

Allcharles replied:
`baw-side-nav`?

atruskie commented on src/app/app.component.html:
```diff
@@ -1,24 +1,45 @@
+    [class.menu-layout]="!fullscreen"
+  >
+    <ng-container *ngIf="fullscreen; else menuLayout">
+      <div id="menu">
```
should this element be a `nav`?

atruskie replied:
additionally this seems like a redundant wrapper container, why not give baw-off-canvas the id and remove the wrapper?

atruskie commented on src/app/app.component.scss:
```diff
@@ -10,105 +10,72 @@
   min-height: 100vh;
 
-  // Fullscreen Layout
-  > .container-fluid {
```
this class is removed but still referred to in the template

Allcharles replied:
container-fluid is a bootstrap class

atruskie commented on src/app/components/shared/header/header.component.html:
```diff
@@ -1,41 +1,54 @@
-      [innerText]="title"
-    ></a>
+<nav class="fixed-top">
+  <div id="navbar" class="navbar navbar-expand-lg navbar-dark">
```
again I'm wondering if we can collapse useless wrapper divs

atruskie commented on src/app/components/shared/header/header.component.html:
```diff
@@ -1,41 +1,54 @@
+            <baw-header-item
+              *ngIf="isNavigableMenuItem(header)"
+              [link]="header"
+            ></baw-header-item>
```
not a comment strictly related to this changeset, but these components seem overly specific

atruskie commented on src/app/components/shared/off-canvas/off-canvas.component.ts:
```diff
@@ -0,0 +1,51 @@
+import { takeUntil } from "rxjs";
+
+@Component({
+  selector: "baw-off-canvas",
```
still think we can have a better name. `baw-menu-container`?

atruskie commented on src/app/interfaces/menusInterfaces.ts:
```diff
@@ -199,12 +195,7 @@ export function menuAction<T extends Omit<MenuAction, "kind">>(
-  | MenuLink
-  | MenuRoute
-  | MenuModalWithoutAction
-  | MenuModal;
```
Hmm  why was `MenuModal` removed from being a valid menu item?

Allcharles replied:
There is no current scenario where i need a type that allows both `MenuModalWithoutAction` and `MenuModal`. Often i need a type which combines `MenuLink|MenuRoute|MenuAction` with either `MenuModalWithoutAction` or `MenuModal`. Potentially i need a better naming scheme. `AnyMenuItem (MenuLink|MenuRoute|MenuAction|MenuModal)`/`AnyMenuItemWithoutModalAction(MenuLink|MenuRoute|MenuAction|MenuModalWithoutAction)`? Currently, I use `AnyMenuItem | MenuModalWithoutAction` or `AnyMenuItem | MenuModal` where needed

atruskie replied:
I guess I just see a menu as a collection of menu items... I'd see no reason at all to ever be more specific about the types.

I don't understand the mutual exclusion for the two modal types. Explain to me on chat? Why can't we have a menu that has multiple modals, of both types?

atruskie reviewed (approved):
Looks fine. Small suggestions made but I think the only real feedback I can give is look for ways to simplify and reduce. 

- remove wrapper divs unless needed
- remove components that are only used once

etc...

atruskie commented on src/app/app.component.ts:
```diff
@@ -27,11 +28,15 @@ export class AppComponent extends withUnsubscribe() implements OnInit {
 
-  public constructor(private config: ConfigService, private title: Title) {
+  public constructor(
+    public menu: MenuService,
```
is this unused?

atruskie commented on src/app/components/shared/menu/action-menu/action-menu.component.ts:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * Action Menu Component.
+ * A menu on the right side of the page displaying actions available
```
```suggestion
 * A menu on located close the current subject of the page, displaying relevant actions available
```

atruskie commented on src/app/components/shared/menu/link/link.component.ts:
```diff
@@ -93,6 +107,21 @@ export class MenuLinkComponent implements OnChanges {
+      .subscribe(({ queryParams, params }) => {
+        this.queryParams = queryParams;
+        this.routeParams = params;
+      });
```
seems like something `SharedActivatedRouteDataService` should be doing?

atruskie commented on src/app/helpers/page/pageComponent.ts:
```diff
@@ -29,6 +32,14 @@ export class PageComponent extends withUnsubscribe() implements IPageComponent {
+  ) {
+    super();
+    this.routeData.setActivatedRoute(route);
+  }
```
yeah see my main comments - this is all totally unrelated the components

atruskie reviewed (commented):
This is probably an asinine thing to mention now, but `fullscreen` is a bit misleading. `fullwidth` layout might be more accurate? Take it or leave it.

Ok, I realise we chose the injection pattern so we wouldn't be filtering router events... but the amount of boilerplate injected in classes now is becoming excessive.

Do we have any alternatives? Are they worth it?

- observing and filtering router events could mean `SharedActivatedRouteDataService` is set its value once in one way
- or going the other way: https://devblogs.microsoft.com/premier-developer/angular-how-to-simplify-components-with-typescript-inheritance/ (using injector to resolve dependencies indirectly)
    - I think this is a bad choice since it completely break the notion of DI

I was indecisive about my above comments, but for every component that I looked at (at least as well as my memory serves me) the `routeData` and `route` injections are not used by the component. They're orthogonal code-wise, even if strictly speaking these things are related to each other via our concept of a page. It ends up just being a lot of noise and ceremony.

atruskie commented on src/app/components/shared/menu/link/link.component.ts:
```diff
@@ -93,6 +107,21 @@ export class MenuLinkComponent implements OnChanges {
+     */
+    this.sharedRoute.activatedRoute$
+      .pipe(
+        map((route): ActivatedRouteSnapshot => route.snapshot),
```
i still think this can be packaged into `SharedActivatedRouteService`.

plus a name would make the intent a little more discernable... `this.sharedRoute.getSnapshot$`

atruskie commented on src/app/services/shared-activated-route/shared-activated-route.service.ts:
```diff
@@ -0,0 +1,131 @@
+    return this._activatedRoute$;
+  }
+
+  public get data(): Observable<Data> {
```
```suggestion
  public get routeData(): Observable<Data> {
```
Maybe? Context seems a little unclear to me. Or just add some documentation?

atruskie commented on src/app/services/shared-activated-route/shared-activated-route.service.ts:
```diff
@@ -0,0 +1,131 @@
+      // Hit max depth, return undefined
+      if (depth === 5) {
+        return undefined;
+      }
```
Why is this needed?

Arbitrary depths in traversals are a huge code smell for me

atruskie commented on src/app/services/shared-activated-route/shared-activated-route.service.ts:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  private observeRouterEvents(router: Router, route: ActivatedRoute): void {
+    const findPageComponent = (
```
for clarity, i'd suggest moving this function outside of the current one

atruskie commented on src/app/services/shared-activated-route/shared-activated-route.service.ts:
```diff
@@ -0,0 +1,131 @@
+      return _route.children
+        .map((child) => findPageComponent(child, depth + 1))
+        .filter((child) => child instanceof ActivatedRoute)
+        .pop();
```
hmmm, `.pop()` mutates the array, maybe `slice(-1)` is a better intent of what you want

I realise that .map and .filter always return new arrays but pop implies you want to mutate the array and keep the state

atruskie reviewed (changes requested):
Ok this is super close. 

I have some concerns about your recursive event filtering.

I think you want something that looks like this:
```
   router.events
      .pipe(
        filter((e): boolean => e instanceof NavigationEnd),
        takeUntil(this.unsubscribe)
        concatAll(expandChildren)
        filter(isPrimaryOutlet)
        first(isPageComponent)
      ).subscribe(...)
```

Also what is the reason for the arbitrary depth traversal limit?

atruskie reviewed (approved):
I'm a little concerned those observables will be re-created each time the properties are fetched, potentially leading to an explosion of observables after a few navigation steps. Maybe initialize them in the constructor for the service? Depends of how they're disposed of later on.

otherwise it's looking really nice.

Allcharles committed:
Fullscreen layout redesign

Major redesign of website layout to better handle menu and fullscreen layouts across a variety of screen sizes
